### ISSUE TITLE
[Develop] Call to a member function process() on string

### ISSUE THREAD

**Issue by harikt** [zendframework/zend-expressive#453]
Earlier we can add a route as 

```
$router->addRoute(new Route('/welcome', 'Hkt\Psr7AssetExample\Action\WelcomeAction', ['GET'], 'hkt/psr7-asset-example:welcome'));
```

Please pay special attention to `WelcomeAction`. The only requirement was it should be got from the container. But after trying to upgrade I noticed this no longer is working. But we need to get the object from the container itself.

```
$router->addRoute(new Route('/welcome', $di->get('Hkt\Psr7AssetExample\Action\WelcomeAction'), ['GET'], 'hkt/psr7-asset-example:welcome'));
```

The error occurs in 

```php
class DispatchMiddleware implements ServerMiddlewareInterface
{   
    public function process(ServerRequestInterface $request, DelegateInterface $delegate)
    {
        $routeResult = $request->getAttribute(RouteResult::class, false);
        if (! $routeResult) {
            return $delegate->process($request);
        }
 
        $middleware = $routeResult->getMatchedMiddleware();
 
        // $middleware obtained from getMatchedMiddleware is a string .

        return $middleware->process($request, $delegate);
    }
}
```

**Comment by xtreamwayz** [comment]
Since you are not using the application itself to setup routes ($app->addRoute, $app->route(), $app->get(), etc.), it looks like you are missing an important call to [MarshalMiddlewareTrait::prepareMiddleware()](https://github.com/zendframework/zend-expressive/blob/develop/src/Application.php#L314) that transforms the Middleware string to [Middleware\LazyLoadingMiddleware](https://github.com/zendframework/zend-expressive/blob/develop/src/MarshalMiddlewareTrait.php#L75). That `LazyLoadingMiddleware` actually gets the middleware from the container.

The preferred way to add routes in expressive 2.0 is:
```php
$app->get('/', App\Action\HomePageAction::class, 'home');
$app->get('/api/ping', App\Action\PingAction::class, 'api.ping');
```

If you use it like that it does all the magic for you.

**Comment by harikt** [comment]
Hm... So do you think adding a Route via the Router is wrong which accepts a string ? I think `DispatchMiddleware` probably need the LazyLoadingMiddleware then.

**Comment by harikt** [comment]
Basically my idea is not to tie it to any framework, but build module that can use psr-7, psr-15 and psr-17 etc. May be it will tie to some Di container though :-( .

**Comment by xtreamwayz** [comment]
I don't know if it's wrong, I'm only saying that the behavior seems to be changed between version 1.0 and 2.0. 

Before this was done in [Application::dispatchMiddleware()](https://github.com/zendframework/zend-expressive/blob/master/src/Application.php#L438), in 2.0 it's done in MarshalMiddlewareTrait::prepareMiddleware() and thus it's the missing link in your code. Note the method names: Before it was done during depatching the middleware and that's why your code worked. In 2.0 it's done while adding routes and preparing middleware and that's why your way doesn't work anymore.

So if you don't want to tie it, grab it from the container itself. However when you do that, it means it will instantiate all routed middleware even when they are not used. That's why 2.0 uses the `LazyLoadingMiddleware` as a wrapper around routed middleware to start it up only when needed.

**Comment by danizord** [comment]
Just for info we ran into the same issue here when using Expressive 2 with config-based routes.

**Comment by xtreamwayz** [comment]
@danizord Can you give more details on this? Looking at the code it should work as expected:

`ApplicationFactory` calls `ApplicationConfigInjectionTrait::injectRoutesFromConfig()` which adds routes directly to `Application::route()`, which calls the `MarshalMiddlewareTrait::prepareMiddleware()` method.

Unless you used the router standalone as well, which would explain the same issue.

**Comment by harikt** [comment]
@xtreamwayz 

> I don't know if it's wrong 

I leave this open to discussion. But this looks wrong to me.

**Comment by danizord** [comment]
@xtreamwayz here is a reduced failing test: https://gist.github.com/danizord/afbc4c0a5d222c4dd5ad5d3120d35a35

**Comment by xtreamwayz** [comment]
@harikt
> So do you think adding a Route via the Router is wrong which accepts a string ?
 
It's not wrong, it's how expressive does it internally. You are just missing a crucial step that grabs the middleware from the container. Expressive does this for you. So if you want an independent router you need to write your own wrapper that does this. 

FastRoute, Aura.router and zend-router are standalone. zend-expressive-router is the bridge between each of those routers and Expressive. In other words it's build specifically for expressive.

@danizord 

It looks like the routes from the config are not prepared. This seems to fix it in `ApplicationConfigInjectionTrait::injectRoutesFromConfig`.

```php
            $name  = isset($spec['name']) ? $spec['name'] : null;

            $middleware = $this->prepareMiddleware(
                $spec['middleware'],
                $this->router,
                $this->responsePrototype,
                $this->container
            );

            $route = new Route($spec['path'], $middleware, $methods, $name);
```

I'm not sure if I can just add this code since it calls another trait which I can't import as it would conflict because it's already imported into Application. So it works in the application, but it would miss the import itself, which might fail tests etc.

@weierophinney Any ideas on how to properly fix it?

**Comment by harikt** [comment]
Hi @xtreamwayz ,

https://github.com/zendframework/zend-expressive-router/blob/ef8453bb4f0dff66b9e4cdeb410f266ba13d6d65/src/RouteResult.php#L150  . The `getMatchedMiddleware` can return any of  `@return false|callable|string|MiddlewareInterface|array` . If it is `string`, `callable` currently this fails. So my whole point is either you need to change the signature or let the `DispatchMiddleware` do the stuff.

Example routes : 

```
$router->addRoute(new Route('/', $di->lazyGet('Hkt\Psr7AssetExample\Action\WelcomeAction'), ['GET'], 'hkt/psr7-asset-example:welcome'));
$router->addRoute(new Route('/', 'Hkt\Psr7AssetExample\Action\WelcomeAction', ['GET'], 'hkt/psr7-asset-example:welcome'));
```

The only way currently this works is 

```
$router->addRoute(new Route('/', $di->lazyGet('Hkt\Psr7AssetExample\Action\WelcomeAction')(), ['GET'], 'hkt/psr7-asset-example:welcome'));
$router->addRoute(new Route('/', $di->get('Hkt\Psr7AssetExample\Action\WelcomeAction'), ['GET'], 'hkt/psr7-asset-example:welcome'));
```

Which is not lazy really.

I will look into how to fix this passing the container to the `DispatchMiddleware` and check if the returned one is callable then call it.

**Comment by harikt** [comment]
I have opened a PR https://github.com/zendframework/zend-expressive/pull/454 which will fix the issue. @danizord can you look if that resolves your issue also? I believe yes.

**Comment by weierophinney** [comment]
Another way to do this is to create `LazyLoadingMiddleware` instances that you pass to your routes:

```php
use Hkt\Psr1AssetExample\Action\WelcomeAction;
use Zend\Diactoros\Response;
use Zend\Expressive\Middleware\LazyLoadingMiddleware;
use Zend\Expressive\Router\Route;

$prototype = new Response();

$router->addRoute(new Route(
    '/',
    new LazyLoadingMiddleware($di, $prototype, WelcomeAction::class),
    ['GET'],
    'hkt/psr7-asset-example:welcome'
));
```

This approach ensures that you have http-interop middleware composed in your router, so that the dispatch middleware can operate as expected without changes.

Alternately, create your own `DispatchMiddleware` implementation:

```php
function (ServerRequestInterface $request, DelegateInterface $delegate) use ($container, $prototype)
{
    $routeResult = $request->getAttribute(RouteResult::class, false);
    if (! $routeResult) {
        return $delegate->process($request);
    }
    
    $middleware = $routeResult->getMatchedMiddleware();
    
    if (! $middleware instanceof MiddlewareInterface) {
        $middleware = new LazyLoadingMiddleware($container, $prototype, $middleware);
    }
    
    return $middleware->process($request);
}
```

I'm still trying to decide if it makes sense for the dispatch middleware to check for non-interop middleware and do this for you. On the one hand, I'd prefer not to, as we want to encourage users to use the `Application` instance to setup routing as much as possible; there's very little reason to directly create the `Route` instances yourself. On the flip side, since the `Route` instances are liberal in what they allow for middleware, we likely should be more defensive in that regard.

I'll review your pull request, and make a decision this coming week.

**Comment by weierophinney** [comment]
Fixed with #455.
